### PR TITLE
add support for converting to WKT to EPSG

### PR DESCRIPTION
**Pull request by alex-s-gardner**

### CODE REVIEW

File: .vscode/settings.json [hunk 1]
```diff
@@ -0,0 +1 @@
+{}
```
Comment (yeesian) [inline .vscode/settings.json:1]: Can we revert the inclusion of this file?

File: src/spatialref.jl [hunk 2]
```diff
@@ -600,6 +600,8 @@ Convert this SRS into a nicely formatted WKT string for display to a person.
 function toWKT(spref::AbstractSpatialRef, simplify::Bool)::String
     wktptr = Ref{Cstring}()
     result = GDAL.osrexporttoprettywkt(spref, wktptr, simplify)
+
+    println(restul)
```
Comment (yeesian) [inline src/spatialref.jl:604]: Please remove print statements

File: src/spatialref.jl [hunk 3]
```diff
@@ -616,6 +618,31 @@ function toPROJ4(spref::AbstractSpatialRef)::String
...
+    result = GDAL.osrgetauthorityname(spref.ptr, "PROJCS")
+    projcs = "PROJCS"
+
+    if isnothing(result)
+        result = GDAL.osrgetauthorityname(spref.ptr, "GEOGCS")
+        projcs = "GEOGCS"
+    end
+
+    if isnothing(result)
+        error("No PROJCS or GEOGCS Authority found")
```
Comment (yeesian) [inline src/spatialref.jl:636]: Would people prefer an empty string, or an error in this case?

**Review by yeesian — CHANGES_REQUESTED** [review]
Thank you!

File: src/spatialref.jl (hunk 3 above)
Comment (alex-s-gardner) [reply src/spatialref.jl:636]: I was thinking the same thing but passing an empty string or "EPSG:" to GFT.EPSG() will throw an error

**Comment by alex-s-gardner** [comment]
a more full featured retrieval of EPSG will require a pyproj like to_epsg:

https://github.com/JuliaGeo/Proj.jl/issues/88

File: src/spatialref.jl (hunk 3 above)
Comment (rafaqz) [reply src/spatialref.jl:636]: `GFT.EPSG` actually holds an integer internally, so the constructor checks that a string is `epsg:NNNN` and parses then Ns.

File: src/spatialref.jl [hunk 4]
```diff
@@ -616,6 +618,31 @@ function toPROJ4(spref::AbstractSpatialRef)::String
...
 end
 
+"""
+    toEPSG(spref::AbstractSpatialRef)
+
+If possible, export EPSG code for this coordinate system.
+"""
+function toEPSG(spref::AbstractSpatialRef)::String
+    result = GDAL.osrgetauthorityname(spref.ptr, "PROJCS")
+    projcs = "PROJCS"
```
Comment (rafaqz) [inline src/spatialref.jl:628]: Its confusing defining `projcs` here - why not as an `else` path of the conditional below?

Comment (alex-s-gardner) [reply src/spatialref.jl:628]: now moved

File: src/spatialref.jl (hunk 3 above)
Comment (alex-s-gardner) [reply src/spatialref.jl:636]: so I think our only options are [1] throw an error or [2] set epsg = 0

File: .vscode/settings.json (hunk 1 above)
Comment (alex-s-gardner) [reply .vscode/settings.json]: done

File: src/spatialref.jl (hunk 3 above)
Comment (yeesian) [reply src/spatialref.jl:636]: I'm okay with keeping the current/proposed behavior of raising an error, so I'll resolve this -- but feel free to re-open or create a ticket for discussion if it remains controversial.

File: src/convert.jl [hunk 5]
```diff
@@ -112,10 +112,16 @@ end
 function unsafe_convertcrs(::Type{<:GFT.ProjString}, crsref)
     return GFT.ProjString(toPROJ4(crsref))
 end
+function unsafe_convertcrs(::Type{<:GFT.EPSG}, crsref)
```
Comment (yeesian) [inline src/convert.jl:115]: Can we add a test for this method? It seems to not be covered per https://app.codecov.io/gh/yeesian/ArchGDAL.jl/pull/391.

**Review by yeesian — DISMISSED** [review]

File: src/convert.jl (hunk 5 above)
Comment (evetion) [reply src/convert.jl:115]: This method is overwritten at line 124, so we need to remove the duplicate.

File: src/spatialref.jl [hunk 6]
```diff
@@ -616,6 +617,31 @@ function toPROJ4(spref::AbstractSpatialRef)::String
     return unsafe_string(projptr[])
 end
 
+"""
+    toEPSG(spref::AbstractSpatialRef)
+
+If possible, export EPSG code for this coordinate system.
+"""
+function toEPSG(spref::AbstractSpatialRef)::String
```
Comment (evetion) [inline src/spatialref.jl:625]: EPSG is a code, so this should return an Int? GFT.EPSG(1) does also work, so it should pose no further changes.

File: src/spatialref.jl [hunk 7]
```diff
@@ -616,6 +617,31 @@ function toPROJ4(spref::AbstractSpatialRef)::String
     return unsafe_string(projptr[])
 end
 
+"""
+    toEPSG(spref::AbstractSpatialRef)
+
+If possible, export EPSG code for this coordinate system.
```
Comment (evetion) [inline src/spatialref.jl:623]:
```suggestion
Export EPSG code for this coordinate system if available.
```

File: src/spatialref.jl [hunk 8]
```diff
@@ -616,6 +617,31 @@ function toPROJ4(spref::AbstractSpatialRef)::String
...
+        projcs = "GEOGCS"
+    end
+
+    if isnothing(result)
+        error("No PROJCS or GEOGCS Authority found")
+    elseif result == "EPSG"
+            epsg = GDAL.osrgetauthoritycode(spref.ptr, projcs)
+        return "EPSG:$(epsg)"
+    else
+        @ogrerr result "is not an EPSG authority"
```
Comment (evetion) [inline src/spatialref.jl:641]: The use of the @ogrerr macro only triggers if GDAL itself failed. Ideally you use this after each call to `osrgetauthoritycode` and `osrgetauthorityname`. Here you probably just want to use error().

**Review by evetion — CHANGES_REQUESTED** [review]

File: src/spatialref.jl (hunk 6 above)
Comment (alex-s-gardner) [reply src/spatialref.jl:625]: I'm not sure what we want to do here... changing this to an Int won't be a prblem for GFT but are there other applicaitons that use crs text?

Comment (evetion) [reply src/spatialref.jl:625]: Well you introduced this function here, so no one uses it yet.

Comment (alex-s-gardner) [reply src/spatialref.jl:625]: Now updated to return an Int32

File: src/convert.jl (hunk 5 above)
Comment (alex-s-gardner) [reply src/convert.jl:115]: now fixed

**Review by evetion — APPROVED** [review]

**Pull request merged**